---

[//]: # "Header"

<p style="text-align: left;  font-size:18pt; LINE-HEIGHT:30px;">
    <span style="float: left">
     Technische Hochschule Ingolstadt<br>
     Prof. Dr. Sören Gröttrup <br>
     Laura Dietl
    </span>
    <span style="float: right;">
       Machine Learning 1<br>
        <span style="float: right;">WS 24/25</span>
    </span>
</p>

[//]: # "Header Aufgabenblatt"
<br>
<p style="text-align: center;  font-size:18pt; LINE-HEIGHT:30px;">
     <span style="font-weight: bold;">Aufgabenblatt 4</span><br>
     Themen: Imputing, Neuronale Netze in PyTorch, Backpropagation<br>
     Abgabetermin: 29.11.2024, 23:59 Uhr <br>
     Punkte: 31
</p>


---

**Namen:** Conrad Christoph, Emelie Hauck

---

# 1. Imputing - Dealing with Missing Values - Ein Weg mit fehlenden Werten umzugehen

Bei einer Variablen in einem Datensatz kann es durchaus vorkommen, dass für einige Beobachtungen die Werte nicht bekannt sind. Zum Beispiel kann es sein, dass ein Kunde in einem Online-Shop sein Profil nicht vollkommen ausfüllt, so dass die Handynummer, das Alter oder gewisse Interessen nicht bekannt sind. Die Variable Handynummer in diesem Datensatz wäre dann für den jeweiligen Kunden nicht befüllt.

Einige Algorithmen des Maschinellen Lernens können mit diesen fehlenden Werten nicht umgehen. Damit man einen Algorithmus trotzdem auf einen Datensatz mit einer Variable mit fehlenden Werten anwenden kann, gibt es einige einfache Strategien.

**Ignoriere die fehlenden Werte (Omission)**

Die einfachste und radikalste Methode ist die fehlenden Werte für das Training zu entfernen. Dies kann man auf zwei Arten machen:
1. Entfernen der Variable aus dem Trainingsdatensatz. Z.B. im obigen Beispiel die Variable Handynummer nicht als Input-Variable zu nehmen.
1. Entfernen aller Zeilen im Datensatz, die einen fehlenden Wert in einer zu berücksichtigen Variable aufweisen. Z.B. im obigen Beispiel alle Kunden ohne Handynummer aus dem Tainingsdatensatz ausschließen.

**Imputing**

Anstatt die fehlenden Werte zu entfernen, kann man auch versuchen, diese mit "geeigneten" Werten aufzufüllen. Dafür gibt es mehrere einfache Methoden.
1. _Constant Value Imputing_ Wenn der Kontext es hergibt, können alle fehlenden Werte einer Variablen mit einem Standardwert aufgefüllt werden. 
    1. Bei einer kategoriellen Variable können z.B. die fehlenden Werte mit 'unknown' oder 'other' ersetzt werden. Wenn z.B. bei einem Kunden der Wert der Variable "Intresse an Sportartikeln" fehlt, kann dieser die neue Ausprägung "unknown" erhalten.
    1. Wenn bei einer numerischen Variablen nur positive Werte auftreten können, wie z.B. bei der Variable "Anzahl an Käufe", kann der fehlende Wert sinnvollerweise mit 0 ersetzt werden.
1. _Mean/Median/Mode Imputing_ Die fehlenden Werte einer Variablen werden durch deren arithmetisches Mittel, den Median oder den Modus ersetzt. Diese repräsentiern "typische" Werte für die Variable.
    1. Bei kategoriellen Variablen werden die fehlenden Werte durch die am häufigsten auftretende Ausprägung ersetzt (den Modus).
    1. Bei numerischen Variablen werden die fehlenden Werte mit dem arithmetischen Mittel oder dem Median der anderen Ausprägungen der Variable ersetzt.

Dies sind nur die einfachen und univariaten Methoden des Imputing. Weitere multivariate Imputing-Verfahren wie die Benutzung von Vorhersagemodellen werden in der Vorlesung besprochen.

## Aufgabe 1.1 (Vor- & Nachteile der Imputing-Verfahren) [6 Punkte]
Betrachten Sie die oben genannten Omission und Imputing-Verfahren. Welche Vor- und Nachteile könnten diese Verfahren jeweils haben?

**Antwort zu Aufgabe 1.1**


Omission:
| Verfahren | Vorteile             | Nachteile                                                                                              |
|-----------|----------------------|--------------------------------------------------------------------------------------------------------|
| 1.        |- Alle Daten sind Uniform <br>- eingefügte Werte ändern die Ergebnisse nicht/keine nicht vorhandenen Daten müssen angenommen werden <br>- alle Datenpunkte bleiben erhalten <br > - unkompliziertes/schnelles Verfahren|- Die entfernte Variable ist potentiell ausschlaggebend für das Ergebnis <br>                             |
| 2.        |- Alle Daten sind Uniform <br>- eingefügte Werte ändern die Ergebnisse nicht/keine nicht vorhandenen Daten müssen angenommen werden <br>- potentiell ausschlaggebende Variablen bleiben erhalten                           |- Sollten viele Datenpunkte unvolsständig sein, so führt das Verfahren zu großem Verlust knapper Daten   |

Imputing:
| Verfahren                      | Vorteile                                                                                                                                     | Nachteile |
|--------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|-----------|
| Constant Value Imputing        |- keine nicht vorhandenen Angaben werden angenommen<br>- Struktur der Daten bleibt erhalten <br>- Stichprobenumfang bleibt erhalten           |- Das Modell könnte Ergebnisse o. zu viel Information aus der Unknown Variable ziehen<br> - Es werden nicht korrekte Daten angenommen, die die Vorhersage verzerren könnten                                                |
| Mean/Median/Mode Imputing      |- kann über verschiedene Annahmemethoden (Mean,Median,etc.) Strukturen im Datensatz berücksichtigen<br>- Stichprobenumfang bleibt erhalten    | - basiert auf Annahmen, die teilweise aufwändig errechnet werden müssen<br>- es könnten systematische Verzerrungen entstehen                                                                                            | 


# 2. Neuronale Netze mit PyTorch

### Daten

Der Datensatz `gt.csv` (zu finden auf der Moodle-Seite des Praktikums (https://moodle.thi.de/course/view.php?id=6824)) ist eine etwas abgewandelte und bearbeitete Version des Datensatzes aus dem UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/datasets/Gas+Turbine+CO+and+NOx+Emission+Data+Set). 

Der Datensatz enthält 36733 Messpunkte von je 11 Sensoren einer Gasturbine im Nordwesten der Türkei. Die einzelnen Werte sind dabei jeweils Aggregate (Summe oder Mittelwert) über eine Stunde. Die Sensordaten diesen dazu die Gasemissionender Turbine, insb. CO und NOx (NO & NO2), zu messen.

#### Initialisierung der Daten
Hier finden Sie bereits geschriebenen Code, der die Daten einliest und in Trainings- und Testdaten splittet. __Bitte verändern Sie diesen Code nicht__, sondern fügen Ihren Code unterhalb ein.

In [ ]:
# Packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split

plt.style.use('ggplot')

In [ ]:
# Load data
file_path = "Daten Blatt 4/gt.csv"
data = pd.read_csv(file_path)
data

In [ ]:
# Define target and feature
y = data.NOX
X = data.drop(columns=["CO", "NOX"])

In [ ]:
# Split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=1)

X_train = pd.DataFrame(X_train, columns=X.columns)
X_test = pd.DataFrame(X_test, columns=X.columns)

## Aufgabe 2.1 (Datenaufbereitung) [4 Punkte]

Untersuchen Sie den Datensatz der Sensordaten einer Gasturbine `gt.csv`.

1. Welche Variablen weisen fehlende Werte auf?
1. Ersetzen Sie diese mittels eines sinnvollen oben genannten Imputing-Verfahrens. _Hinweis_: Sie können dafür die Funktion `SimpleImputer()` aus dem `sklearn` Paket verwenden. Informieren Sie sich vorher über deren Funktionsweise und Optionen.
1. Was für ein Vorverarbeitungsschritt scheint noch sinnvoll zu sein? Führen Sie diesen ggf. auch durch.

_Achtung_: Behalten Sie bei Ihren Datenoperationen die gegebene Aufteilung in Training- und Testdaten bei.

### Lösung Aufgabe 2.1

In [ ]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

In [ ]:
# 2.1.1

empty_cells = data.isnull().any(axis=0)
print(empty_cells)

Die Variablen TAT & CDP weisen fehlende Werte auf.

In [ ]:
# 2.1.2

imputer = SimpleImputer(strategy='median')

imputer.fit(X_train)
imputed_x_train = imputer.transform(X_train)

imputer.fit(X_test)
imputed_x_test = imputer.transform(X_test)

#### 2.1.3

Die Daten weisen unterschiedliche Größen auf (z.B. AT im 1-2 stelligen Bereich vor dem Komma und AP im vierstelligen Bereich vor dem Komma) entsprechend wäre eine Skalierung sinnvoll, da einige Verfahren nur mit skalierten Daten präzise Ergebnisse liefern.    

In [ ]:
scaler = StandardScaler()

x_train_sc_imp = scaler.fit_transform(imputed_x_train)

x_test_sc_imp = scaler.transform(imputed_x_test)

## Aufgabe 2.2 (Training in PyTorch) [12 Punkte]
Erstellen Sie ein Feedforward Neuronales Netz auf dem in der vorherigen Aufgabe 2.1 aufbereiteten Datensatz mittels PyTorch, welches basierend auf den anderen Sensordaten die NOx-Emission vorhersagen kann. Benutzen Sie dabei den L2-Loss als Kostenfunktion. Welchen RMSE erzielt Ihr trainiertes Neuronales Netz, wenn Sie dies auf die Testdaten (`X_test, y_test`) anwenden?

In dieser Aufgabe gibt es keinerlei weitere Vorgabe an das zu implementierende Modell. Sie sind in der Struktur des Netzes und dem Vorgehen frei. Um aber die volle Punkzahl zu erhalten, müssen Sie ein sinnvolles Netz trainieren, d.h. die Optimierung mittels Gradient Descent konvergiert, kein Overfitting etc.. Versuchen Sie auch den RMSE so klein wie möglich zu bekommen.

_Hinweis_: Anstatt Ihre aufbereiteten Daten zu nehmen, können Sie auch schon vorverarbeitete Daten einlesen durch den Befehl `data = torch.load("Daten Blatt 4/gt_prepared")`
Dieser liest ein Dictionary ein, in welchem Sie Tensoren der Trainings- und Testdaten finden. Der Befehl `data.keys()` zeigt Ihnen die Key-Namen. 

### Lösung der Aufgabe 2.2

In [ ]:
# Load packages
import torch
import torch.nn as net
import torch.nn.functional as F
import torch.optim as optim

In [14]:
data = torch.load("Daten Blatt 4/gt_prepared")
data.keys()

X_train_new = data["train_data"]
y_train_new = data["train_target"]
y_train_new = y_train_new.unsqueeze(1)

X_test_new = data["test_data"]
y_test_new = data["test_target"]

/tmp/ipykernel_2003/815055729.py:1: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  data = torch.load("Daten Blatt 4/gt_prepared")


In [16]:
class FeedforwardNeuralNet(net.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(FeedforwardNeuralNet, self).__init__()
        self.neuralnet = net.Sequential(
            net.Linear(input_size, hidden_size),
            net.ReLU(),
            net.Linear(hidden_size,32),
            net.ReLU(),
            net.Linear(32,output_size)
        ) 
    
    def forward(self, x):
        return self.neuralnet(x)
    
if len(X_train_new.shape) == 1:
    X_train_new = X_train_new.unsqueeze(1)
if len(X_test_new.shape) == 1:
    X_test_new = X_test_new.unsqueeze(1)

input_size = X_train_new.shape[1]
hidden_size = 64
output_size = 1
alpha = 0.2
epochs = 80

model = FeedforwardNeuralNet(input_size, hidden_size, output_size)

loss = net.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=alpha)

for epoch in range(epochs):
    optimizer.zero_grad()
    y_pred = model(X_train_new)

    train_loss = F.mse_loss(y_pred, y_train_new)
    train_loss.backward()
    optimizer.step()
    print(f"Epoch {epoch+1}/{epochs}, Loss: {train_loss.item()}")



Epoch 1/80, Loss: 4384.796875
Epoch 2/80, Loss: 673449.3125
Epoch 3/80, Loss: 115071.3828125
Epoch 4/80, Loss: 1.7596517944307577e+21
Epoch 5/80, Loss: 8.489572929835211e+19
Epoch 6/80, Loss: 3.056244741812676e+19
Epoch 7/80, Loss: 1.1002486304200983e+19
Epoch 8/80, Loss: 3.960896630818865e+18
Epoch 9/80, Loss: 1.4259232323970007e+18
Epoch 10/80, Loss: 5.133322578349261e+17
Epoch 11/80, Loss: 1.8479959220473037e+17
Epoch 12/80, Loss: 6.652787518393549e+16
Epoch 13/80, Loss: 2.395003472261939e+16
Epoch 14/80, Loss: 8622015850217472.0
Epoch 15/80, Loss: 3103925083308032.0
Epoch 16/80, Loss: 1117412850139136.0
Epoch 17/80, Loss: 402268717318144.0
Epoch 18/80, Loss: 144816734208000.0
Epoch 19/80, Loss: 52134011731968.0
Epoch 20/80, Loss: 18768245817344.0
Epoch 21/80, Loss: 6756568137728.0
Epoch 22/80, Loss: 2432364969984.0
Epoch 23/80, Loss: 875651268608.0
Epoch 24/80, Loss: 315234418688.0
Epoch 25/80, Loss: 113484406784.0
Epoch 26/80, Loss: 40854384640.0
Epoch 27/80, Loss: 14707581952.0
E

# 3. Parameter eines Neuronalen Netzes

## Aufgabe 3.1 [3 Punkte]
Gegeben ist das in PyTorch definierte Neuronale Netze
   
![fc_nn](Bilder/fc_nn.png)

Bestimmen Sie  die Anzahl zu optimierender Parameter.

### Antwort zu Aufgabe 3.1

...


# 4. Backpropagation

## Aufgabe 4.1 [6 Punkte]
Gegeben ist folgendes Neuronale Netze
   
![fc_nn](Bilder/NN_backpropagation.png)

mit folgenden Gewichtsmatrizen 

$$W^{(1)}=\left(\begin{array}{cc} 
1 & 1\\ 
2 & -2
\end{array}\right)
\quad\text{und}\quad
W^{(2)}=\left(\begin{array}{c} 
1\\ 
-1
\end{array}\right) $$

und Biases

$$b^{(1)}=\left(\begin{array}{c} 
-1\\ 
0
\end{array}\right)
\quad\text{und}\quad
b^{(2)}= -4$$

Die Aktivierungsfunktionen sind folgende:
* Hidden Layer: ReLU
* Output Layer: Identität

Als Verlustfunktion wird der L2-Loss genommen
$$L(y,a)=\frac12(y-a)^2$$

__Aufgabe:__
1. Führen Sie ein Update der Parameter mittels Backpropagation durch für den Datenpunkt
$$x=\left(\begin{array}{c} 
1\\ 
1 
\end{array}\right) $$
mit wahrem Zielwert 
$$ y= -1$$
und der Lernrate $\alpha=0.1$ durch.
1. Wie groß ist der neue Verlust nach dem Parameterupdate? Vergleichen Sie das mit dem Verlust vor durchführung des Updates der Parameter.


### Lösung Aufgabe 4.1

...